# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/QuotaCommon.cpp
# Commit: dc0c88b81dc2
# Full Hash: dc0c88b81dc26894c34ae58fb304dc349dd062f0
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2020-12-05 09:38:58
# Regressor Bug: 1678030
# File Overlap Count: 1
# Description:
#   Bug 1678030 - Generalize CreateAndExecuteSingleStepStatement to accept an optional bind functor. r=dom-workers-and-storage-reviewers,ttung
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D97462
# ==============================================================================

diff -r 2fb182664097 -r dc0c88b81dc2 dom/quota/QuotaCommon.cpp
--- a/dom/quota/QuotaCommon.cpp	Fri Dec 04 15:10:07 2020 +0000
+++ b/dom/quota/QuotaCommon.cpp	Fri Dec 04 15:15:31 2020 +0000
@@ -157,6 +157,39 @@
   return resultFile;
 }
 
+Result<nsCOMPtr<mozIStorageStatement>, nsresult> CreateStatement(
+    mozIStorageConnection& aConnection, const nsACString& aStatementString) {
+  QM_TRY_RETURN(MOZ_TO_RESULT_INVOKE_TYPED(nsCOMPtr<mozIStorageStatement>,
+                                           aConnection, CreateStatement,
+                                           aStatementString));
+}
+
+template <SingleStepResult ResultHandling>
+Result<SingleStepSuccessType<ResultHandling>, nsresult> ExecuteSingleStep(
+    nsCOMPtr<mozIStorageStatement>&& aStatement) {
+  QM_TRY_INSPECT(const bool& hasResult,
+                 MOZ_TO_RESULT_INVOKE(aStatement, ExecuteStep));
+
+  if constexpr (ResultHandling == SingleStepResult::AssertHasResult) {
+    MOZ_ASSERT(hasResult);
+    (void)hasResult;
+
+    return WrapNotNullUnchecked(std::move(aStatement));
+  } else {
+    return hasResult ? std::move(aStatement) : nullptr;
+  }
+}
+
+template Result<SingleStepSuccessType<SingleStepResult::AssertHasResult>,
+                nsresult>
+ExecuteSingleStep<SingleStepResult::AssertHasResult>(
+    nsCOMPtr<mozIStorageStatement>&&);
+
+template Result<SingleStepSuccessType<SingleStepResult::ReturnNullIfNoResult>,
+                nsresult>
+ExecuteSingleStep<SingleStepResult::ReturnNullIfNoResult>(
+    nsCOMPtr<mozIStorageStatement>&&);
+
 template <SingleStepResult ResultHandling>
 Result<SingleStepSuccessType<ResultHandling>, nsresult>
 CreateAndExecuteSingleStepStatement(mozIStorageConnection& aConnection,
@@ -165,16 +198,7 @@
                                nsCOMPtr<mozIStorageStatement>, aConnection,
                                CreateStatement, aStatementString));
 
-  QM_TRY_UNWRAP(const DebugOnly<bool> hasResult,
-                MOZ_TO_RESULT_INVOKE(stmt, ExecuteStep));
-
-  if constexpr (ResultHandling == SingleStepResult::AssertHasResult) {
-    MOZ_ASSERT(hasResult);
-
-    return WrapNotNullUnchecked(stmt);
-  } else {
-    return hasResult ? stmt : nullptr;
-  }
+  return ExecuteSingleStep<ResultHandling>(std::move(stmt));
 }
 
 template Result<SingleStepSuccessType<SingleStepResult::AssertHasResult>,